File: DONGUK/프로그래머스_WEEK_ETC/양궁대회.py
import copy

maxDiff = 0
candidates = []

def caclResult(apeach, ryan):
    apeach_score = 0
    ryan_score = 0
    for i in range(11):
        score = 10 - i
        if apeach[i] == 0 and ryan[i] == 0:
            continue
            
        if apeach[i] >= ryan[i]:
            apeach_score += score
        
        else:
            ryan_score += score
        
    return ryan_score - apeach_score


def dfs(ryan, i, info, n):
    global maxDiff, candidates
    
    if i == 11:
        if n > 0:
            ryan[i-1] = n
        diff = caclResult(info, ryan)
        if diff <= 0:
            return
        
        res = copy.deepcopy(ryan)
        if maxDiff < diff:
            maxDiff = diff
            candidates = [res]
            return
        
        if maxDiff == diff:
            candidates.append(res)
            
        return
    
    if info[i] < n:
        ryan.append(info[i] + 1)
        dfs(ryan, i+1, info, n-info[i]-1)
        ryan.pop()
    
    ryan.append(0)
    dfs(ryan, i+1, info, n)
    ryan.pop()
    

def solution(n, info):
    answer = []
    
    dfs([], 0, info, n)
        
    if not candidates:
        answer.append(-1)
    else:
        candidates.sort(key=lambda x: x[::-1], reverse=True)
        answer = candidates[0]
        
    return answer

File: JUNHEE/프로그래머스_옛날문제/양궁대회.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;
/*
    자.. dfs로 푸는데, 각 칸을 먹은 경우, 안먹은 경우로 가자..
*/
vector<int> result;
int t_max = -1;

void calc(vector<int> apeach, vector<int> ryan){
    int a = 0, r = 0;
    for(int i = 0; i < 11; i++){
        if(apeach[i] < ryan[i])
            r += 10 - i;
        else if(0 < apeach[i])
            a += 10 - i;
    }
    if(r - a > 0 && t_max < (r - a)){
        cout << r << " " << a << endl;
        result = ryan;
        t_max = r - a;
    }
    else if(r - a > 0 && t_max == (r - a)){
        for(int i = 10; i >=0; i--){
            if(result[i] != 0 || ryan[i] != 0){
                if(ryan[i] > result[i])
                    result = ryan;
                else if(ryan[i] == result[i])
                    continue;
                else
                    break;
            }
        }
    }
}

void dfs(int idx, int arrow, vector<int> apeach, vector<int>ryan){
    if(idx == 11 || arrow == 0){
        if(arrow != 0){
            ryan[10] += arrow;
        }
        calc(apeach, ryan);
        return;
    }
    if(apeach[idx] < arrow){
        ryan[idx] += apeach[idx] + 1;
        arrow -= apeach[idx] + 1;
        dfs(idx + 1, arrow, apeach, ryan);
        arrow += apeach[idx] + 1;
        ryan[idx] -= apeach[idx] + 1;
    }
    dfs(idx + 1, arrow, apeach, ryan);
    
}

vector<int> solution(int n, vector<int> info) {
    vector<int> answer(1, -1);
    vector<int> ryan(11,0);
    result.resize(11,0);
    dfs(0, n, info, ryan);
    for(auto i : result){
        if(i != 0){
            answer = result;
        }
    }
    return answer;
}
